test/Npgsql.Tests/DataSourceTests.cs

```diff
@@ -380,4 +381,86 @@ public async Task ConfigureJsonOptions_is_order_independent()
             Assert.That(reader.GetFieldValue<Test>(0).Id, Is.EqualTo(1));
         }
     }
+
+    [Test]
+    public async Task ReloadTypes()
+    {
+        using var adminConnection =  OpenConnection();
+        var type = await GetTempTypeName(adminConnection);
+
+        using var dataSource = CreateDataSourceWithType(type);
+        using var connection = dataSource.OpenConnection();
+
+        connection.ExecuteNonQuery($"CREATE TYPE {type} AS ENUM ('sad', 'ok', 'happy')");
+        dataSource.ReloadTypes();
+
+        Assert.Throws<InvalidCastException>(() => connection.ExecuteScalar($"SELECT 'happy'::{type}"));
+
+        // Close connection and reopen to make sure it picks up the new type and mapping from the data source
+        connection.Close();
+        connection.Open();
+
+        Assert.DoesNotThrow(() => connection.ExecuteScalar($"SELECT 'happy'::{type}"));
+    }
+
+    [Test]
+    public async Task ReloadTypesAsync()
+    {
+        await using var adminConnection = await OpenConnectionAsync();
+        var type = await GetTempTypeName(adminConnection);
+
+        await using var dataSource = CreateDataSourceWithType(type);
+        await using var connection = await dataSource.OpenConnectionAsync();
+
+        await connection.ExecuteNonQueryAsync($"CREATE TYPE {type} AS ENUM ('sad', 'ok', 'happy')");
+        await dataSource.ReloadTypesAsync();
+
+        Assert.ThrowsAsync<InvalidCastException>(async () => await connection.ExecuteScalarAsync($"SELECT 'happy'::{type}"));
+
+        // Close connection and reopen to make sure it picks up the new type and mapping from the data source
+        await connection.CloseAsync();
+        await connection.OpenAsync();
+
+        Assert.DoesNotThrowAsync(async () => await connection.ExecuteScalarAsync($"SELECT 'happy'::{type}"));
+    }
+
+    [Test]
+    public async Task ReloadTypesAsync_across_data_sources()
+    {
+        await using var adminConnection = await OpenConnectionAsync();
```

Comment:
I think no need for an adminConnection here, strictly speaking - we can just use one of the data sources.

Reply:
I do not see how to use the data source to get the temp type as we need to map it on its own builder.
@roji Feel free to improve it as part of this PR.

Reply:
You're right, missed that dependency.